File: ChartsDemo/Classes/Demos/BubbleChartViewController.m
//
//  BubbleChartViewController.m
//  ChartsDemo
//
//  Bubble chart implementation:
//    Copyright 2015 Pierre-Marc Airoldi
//    Licensed under Apache License 2.0
//
//  https://github.com/danielgindi/ios-charts
//

#import "BubbleChartViewController.h"
#import "ChartsDemo-Swift.h"

@interface BubbleChartViewController () <ChartViewDelegate>

@property (nonatomic, strong) IBOutlet BubbleChartView *chartView;
@property (nonatomic, strong) IBOutlet UISlider *sliderX;
@property (nonatomic, strong) IBOutlet UISlider *sliderY;
@property (nonatomic, strong) IBOutlet UITextField *sliderTextX;
@property (nonatomic, strong) IBOutlet UITextField *sliderTextY;

@end

@implementation BubbleChartViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = @"Bubble Chart";
    
    self.options = @[
                     @{@"key": @"toggleValues", @"label": @"Toggle Values"},
                     @{@"key": @"toggleHighlight", @"label": @"Toggle Highlight"},
                     @{@"key": @"toggleStartZero", @"label": @"Toggle StartZero"},
                     @{@"key": @"animateX", @"label": @"Animate X"},
                     @{@"key": @"animateY", @"label": @"Animate Y"},
                     @{@"key": @"animateXY", @"label": @"Animate XY"},
                     @{@"key": @"saveToGallery", @"label": @"Save to Camera Roll"},
                     @{@"key": @"togglePinchZoom", @"label": @"Toggle PinchZoom"},
                     @{@"key": @"toggleAutoScaleMinMax", @"label": @"Toggle auto scale min/max"},
                     ];
    
    _chartView.delegate = self;
    
    _chartView.descriptionText = @"";
    _chartView.noDataTextDescription = @"You need to provide data for the chart.";
    
    _chartView.drawGridBackgroundEnabled = NO;
    _chartView.dragEnabled = YES;
    [_chartView setScaleEnabled:YES];
    _chartView.maxVisibleValueCount = 200;
    _chartView.pinchZoomEnabled = YES;
    _chartView.leftAxis.startAtZeroEnabled = NO;
    _chartView.rightAxis.startAtZeroEnabled = NO;
    
    ChartLegend *l = _chartView.legend;
    l.position = ChartLegendPositionRightOfChart;
    l.font = [UIFont fontWithName:@"HelveticaNeue-Light" size:10.f];
    
    ChartYAxis *yl = _chartView.leftAxis;
    yl.labelFont = [UIFont fontWithName:@"HelveticaNeue-Light" size:10.f];
    yl.spaceTop = 0.3;
    yl.startAtZeroEnabled = NO;
    yl.spaceBottom = 0.3;

    _chartView.rightAxis.enabled = NO;
    
    ChartXAxis *xl = _chartView.xAxis;
    xl.labelPosition = XAxisLabelPositionBottom;
    xl.labelFont = [UIFont fontWithName:@"HelveticaNeue-Light" size:10.f];
    
    _sliderX.value = 5.0;
    _sliderY.value = 50.0;
    [self slidersValueChanged:nil];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)setDataCount:(int)count range:(double)range
{
    NSMutableArray *xVals = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < count; i++)
    {
        [xVals addObject:[@(i) stringValue]];
    }
    
    NSMutableArray *yVals1 = [[NSMutableArray alloc] init];
    NSMutableArray *yVals2 = [[NSMutableArray alloc] init];
    NSMutableArray *yVals3 = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < count; i++)
    {
        double val = (double) (arc4random_uniform(range));
        double size = (double) (arc4random_uniform(range));
        [yVals1 addObject:[[BubbleChartDataEntry alloc] initWithXIndex:i value:val size:size]];
        
        val = (double) (arc4random_uniform(range));
        size = (double) (arc4random_uniform(range));
        [yVals2 addObject:[[BubbleChartDataEntry alloc] initWithXIndex:i value:val size:size]];
        
        val = (double) (arc4random_uniform(range));
        size = (double) (arc4random_uniform(range));
        [yVals3 addObject:[[BubbleChartDataEntry alloc] initWithXIndex:i value:val size:size]];
    }
    
    BubbleChartDataSet *set1 = [[BubbleChartDataSet alloc] initWithYVals:yVals1 label:@"DS 1"];
    [set1 setColor:ChartColorTemplates.colorful[0] alpha:0.50f];
    [set1 setDrawValuesEnabled:YES];
    BubbleChartDataSet *set2 = [[BubbleChartDataSet alloc] initWithYVals:yVals2 label:@"DS 2"];
    [set2 setColor:ChartColorTemplates.colorful[1] alpha:0.50f];
    [set2 setDrawValuesEnabled:YES];
    BubbleChartDataSet *set3 = [[BubbleChartDataSet alloc] initWithYVals:yVals3 label:@"DS 3"];
    [set3 setColor:ChartColorTemplates.colorful[2] alpha:0.50f];
    [set3 setDrawValuesEnabled:YES];
    
    NSMutableArray *dataSets = [[NSMutableArray alloc] init];
    [dataSets addObject:set1];
    [dataSets addObject:set2];
    [dataSets addObject:set3];
    
    BubbleChartData *data = [[BubbleChartData alloc] initWithXVals:xVals dataSets:dataSets];
    [data setValueFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:7.f]];
    [data setHighlightCircleWidth: 1.5];
    [data setValueTextColor:UIColor.whiteColor];
    
    _chartView.data = data;
}

- (void)optionTapped:(NSString *)key
{
    if ([key isEqualToString:@"toggleValues"])
    {
        for (id<IChartDataSet> set in _chartView.data.dataSets)
        {
            set.drawValuesEnabled = !set.isDrawValuesEnabled;
        }
        
        [_chartView setNeedsDisplay];
    }
    
    if ([key isEqualToString:@"toggleFilled"])
    {
        for (id<ILineChartDataSet> set in _chartView.data.dataSets)
        {
            set.drawFilledEnabled = !set.isDrawFilledEnabled;
        }
        
        [_chartView setNeedsDisplay];
    }
    
    if ([key isEqualToString:@"toggleCircles"])
    {
        for (id<ILineChartDataSet> set in _chartView.data.dataSets)
        {
            set.drawCirclesEnabled = !set.isDrawCirclesEnabled;
        }
        
        [_chartView setNeedsDisplay];
    }
    
    if ([key isEqualToString:@"toggleCubic"])
    {
        for (id<ILineChartDataSet> set in _chartView.data.dataSets)
        {
            set.drawCubicEnabled = !set.isDrawCubicEnabled;
        }
        
        [_chartView setNeedsDisplay];
    }
    
    if ([key isEqualToString:@"toggleHighlight"])
    {
        _chartView.data.highlightEnabled = !_chartView.data.isHighlightEnabled;
        [_chartView setNeedsDisplay];
    }
    
    if ([key isEqualToString:@"toggleStartZero"])
    {
        _chartView.leftAxis.startAtZeroEnabled = !_chartView.leftAxis.isStartAtZeroEnabled;
        _chartView.rightAxis.startAtZeroEnabled = !_chartView.rightAxis.isStartAtZeroEnabled;
        
        [_chartView notifyDataSetChanged];
    }
    
    if ([key isEqualToString:@"animateX"])
    {
        [_chartView animateWithXAxisDuration:3.0];
    }
    
    if ([key isEqualToString:@"animateY"])
    {
        [_chartView animateWithYAxisDuration:3.0];
    }
    
    if ([key isEqualToString:@"animateXY"])
    {
        [_chartView animateWithXAxisDuration:3.0 yAxisDuration:3.0];
    }
    
    if ([key isEqualToString:@"saveToGallery"])
    {
        [_chartView saveToCameraRoll];
    }
    
    if ([key isEqualToString:@"togglePinchZoom"])
    {
        _chartView.pinchZoomEnabled = !_chartView.isPinchZoomEnabled;
        
        [_chartView setNeedsDisplay];
    }
    
    if ([key isEqualToString:@"toggleAutoScaleMinMax"])
    {
        _chartView.autoScaleMinMaxEnabled = !_chartView.isAutoScaleMinMaxEnabled;
        [_chartView notifyDataSetChanged];
    }
}

#pragma mark - Actions

- (IBAction)slidersValueChanged:(id)sender
{
    _sliderTextX.text = [@((int)_sliderX.value + 1) stringValue];
    _sliderTextY.text = [@((int)_sliderY.value) stringValue];
    
    [self setDataCount:(_sliderX.value + 1) range:_sliderY.value];
}

#pragma mark - ChartViewDelegate

- (void)chartValueSelected:(ChartViewBase * __nonnull)chartView entry:(ChartDataEntry * __nonnull )entry dataSetIndex:(NSInteger)dataSetIndex highlight:(ChartHighlight * __nonnull)highlight
{
    NSLog(@"chartValueSelected");
}

- (void)chartValueNothingSelected:(ChartViewBase * __nonnull)chartView
{
    NSLog(@"chartValueNothingSelected");
}

@end

File: ChartsSwiftDemo/Ratings/BubbleChartViewController.swift
//
//  BubbleChartViewController.swift
//  Ratings
//
//  Created by Nelson Tam on 2015-11-15.
//  Copyright © 2015 Nelson Tam. All rights reserved.
//

import Foundation
import Charts

class BubbleChartViewController : DemoBaseViewController, ChartViewDelegate {
    
    @IBOutlet var _chartView: BubbleChartView!
    @IBOutlet var _sliderX: UISlider!
    @IBOutlet var _sliderY: UISlider!
    @IBOutlet var _sliderTextX: UITextField!
    @IBOutlet var _sliderTextY: UITextField!
    @IBOutlet var _options: UIButton!
    
    var options : Array<NSDictionary> = Array<NSDictionary>()
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    func setupMenu() {
        options = [
            ["key": "toggleValues", "label": "Toggle Values"],
            ["key": "toggleHighlight", "label": "Toggle Highlight"],
            ["key": "toggleStartZero", "label": "Toggle StartZero"],
            ["key": "animateX", "label": "Animate X"],
            ["key": "animateY", "label": "Animate Y"],
            ["key": "animateXY", "label": "Animate XY"],
            ["key": "saveToGallery", "label": "Save to Camera Roll"],
            ["key": "togglePinchZoom", "label": "Toggle PinchZoom"],
            ["key": "toggleAutoScaleMinMax", "label": "Toggle auto scale min/max"]
        ]
        var optionNames : Array<String> = Array<String>()
        options.map{ a in optionNames.append(String(a["label"]!))}
        
        self.navigationController?.navigationBar.translucent = false
        self.navigationController?.navigationBar.barTintColor = UIColor(red: 0.0/255.0, green:180/255.0, blue:220/255.0, alpha: 1.0)
        self.navigationController?.navigationBar.titleTextAttributes = [NSForegroundColorAttributeName: UIColor.whiteColor()]
        
        let menuView = BTNavigationDropdownMenu(title: optionNames.first!, items: optionNames)
        menuView.cellHeight = 50
        menuView.cellBackgroundColor = self.navigationController?.navigationBar.barTintColor
        menuView.cellSelectionColor = UIColor(red: 0.0/255.0, green:160.0/255.0, blue:195.0/255.0, alpha: 1.0)
        menuView.cellTextLabelColor = UIColor.whiteColor()
        menuView.cellTextLabelFont = UIFont(name: "Avenir-Heavy", size: 17)
        menuView.arrowPadding = 15
        menuView.animationDuration = 0.5
        menuView.maskBackgroundColor = UIColor.blackColor()
        menuView.maskBackgroundOpacity = 0.3
        menuView.didSelectItemAtIndexHandler = {(indexPath: Int) -> () in
            print("Did select item at index: \(indexPath)")
            self.optionTapped( String(self.options[indexPath]["key"]!))
        }
        
        self.navigationItem.titleView = menuView
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupMenu()
        
        self.title = "Bubble Chart"
        
        _chartView.delegate = self
        
        _chartView.descriptionText = ""
        _chartView.noDataTextDescription = "You need to provide data for the chart."
        
        _chartView.drawGridBackgroundEnabled = false
        _chartView.dragEnabled = false
        _chartView.setScaleEnabled(true)
        _chartView.maxVisibleValueCount = 200
        _chartView.pinchZoomEnabled = true
        _chartView.leftAxis.startAtZeroEnabled = false
        _chartView.rightAxis.startAtZeroEnabled = false
        
        let l: ChartLegend = _chartView.legend
        l.position = ChartLegend.ChartLegendPosition.RightOfChart
        l.font = UIFont(name: "HelveticaNeue-Light", size:10)!
        
        let yl: ChartYAxis = _chartView.leftAxis
        yl.labelFont = UIFont(name: "HelveticaNeue-Light", size:10)!
        yl.spaceTop = 0.3
        yl.startAtZeroEnabled = false
        yl.spaceBottom = 0.3
        
        _chartView.rightAxis.enabled = false
        
        let xl: ChartXAxis = _chartView.xAxis
        xl.labelPosition = ChartXAxis.XAxisLabelPosition.Bottom
        xl.labelFont = UIFont(name: "HelveticaNeue-Light", size:10)!
        
        _sliderX.value = 5.0
        _sliderY.value = 50.0
        self.setDataCount(Int(_sliderX.value) + 1, range:Double(_sliderY.value))
    }
    
    override func didReceiveMemoryWarning()
    {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        if (tableView == _optionsTableView)
        {
            return self.options.count;
        }
        
        return 0;
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCellWithIdentifier("Cell")
        if ((cell == nil)) {
            
            cell = UITableViewCell(style: UITableViewCellStyle.Default, reuseIdentifier: "Cell")
            cell!.backgroundView = nil;
            cell!.backgroundColor = UIColor.clearColor()
            cell!.textLabel!.textColor = UIColor.whiteColor()
        }
        
        cell!.textLabel!.text = String(self.options[indexPath.row]["label"])
        
        return cell!
    }
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        if (tableView == _optionsTableView)
        {
            tableView.deselectRowAtIndexPath(indexPath, animated:true)
            
            if ((_optionsTableView) != nil)
            {
                _optionsTableView!.removeFromSuperview()
                _optionsTableView = nil;
            }
            
            self.optionTapped(String(self.options[indexPath.row]["key"]))
        }
    }
    
    func setDataCount(count: Int, range: Double) {
        var xVals: Array = Array<Int>()
    
        for (var i: Int = 0; i < count; i++)
        {
            xVals.append(i)
        }
    
        var yVals1 : Array = Array<BubbleChartDataEntry>()
        var yVals2 : Array = Array<BubbleChartDataEntry>()
        var yVals3 : Array = Array<BubbleChartDataEntry>()
    
        for (var i: Int = 0; i < count; i++)
        {
            var val: Double = Double(arc4random_uniform(UInt32(range)))
            var size: CGFloat = CGFloat(arc4random_uniform(UInt32(range)))
            yVals1.append(BubbleChartDataEntry(xIndex: i, value: val, size: size))
            
            val = Double(arc4random_uniform(UInt32(range)))
            size = CGFloat(arc4random_uniform(UInt32(range)))
            yVals2.append(BubbleChartDataEntry(xIndex: i, value: val, size: size))

            val = Double(arc4random_uniform(UInt32(range)))
            size = CGFloat(arc4random_uniform(UInt32(range)))
            yVals3.append(BubbleChartDataEntry(xIndex: i, value: val, size: size))
        }
    
        let set1: BubbleChartDataSet = BubbleChartDataSet(yVals: yVals1, label: "DS 1")
        set1.setColor(ChartColorTemplates.colorful()[0], alpha:0.50)
        set1.drawValuesEnabled = true
        let set2: BubbleChartDataSet = BubbleChartDataSet(yVals: yVals2, label: "DS 2")
        set2.setColor(ChartColorTemplates.colorful()[1], alpha:0.50)
        set2.drawValuesEnabled = true
        let set3 : BubbleChartDataSet = BubbleChartDataSet(yVals: yVals3, label: "DS 3")
        set3.setColor(ChartColorTemplates.colorful()[2], alpha:0.50)
        set3.drawValuesEnabled = true
    
        var dataSets : Array = Array<BubbleChartDataSet>()
        dataSets.append(set1)
        dataSets.append(set2)
        dataSets.append(set3)
    
        let data : BubbleChartData = BubbleChartData(xVals: xVals, dataSets:dataSets)
        data.setValueFont(UIFont(name: "HelveticaNeue-Light", size: 7.0))
        data.setHighlightCircleWidth(1.5)
        data.setValueTextColor(UIColor.whiteColor())
    
        _chartView.data = data;
    }

    @IBAction func optionsButtonPressed(sender: UIButton) {
        
    }
    
    override func optionTapped(key: String) {
        if (key == "toggleValues") {
            for set in _chartView.data!.dataSets {
                set.drawValuesEnabled = !set.isDrawValuesEnabled
            }
            _chartView.setNeedsDisplay()
        }
    
        if (key == "toggleFilled") {
            for set in _chartView.data!.dataSets {
                if (set is LineRadarChartDataSet) {
                    let finalSet : LineRadarChartDataSet = set as! LineRadarChartDataSet
                    finalSet.drawFilledEnabled = !(finalSet.drawFilledEnabled)
                }
            }
            _chartView.setNeedsDisplay()
        }
    
        if (key == "toggleCircles") {
            for set in _chartView.data!.dataSets {
                if (set is LineRadarChartDataSet) {
                    let finalSet : LineChartDataSet = set as! LineChartDataSet
                    finalSet.drawCirclesEnabled = !(finalSet.drawCirclesEnabled)
                }
            }
    
            _chartView.setNeedsDisplay()
        }
    
        if (key == "toggleCubic") {
            for set in _chartView.data!.dataSets {
                if (set is LineRadarChartDataSet) {
                    let finalSet : LineChartDataSet = set as! LineChartDataSet
                    finalSet.drawCubicEnabled = !(finalSet.drawCubicEnabled)
                }
            }
    
            _chartView.setNeedsDisplay()
        }
    
        if (key == "toggleHighlight") {
            _chartView.data!.highlightEnabled = !_chartView.data!.isHighlightEnabled
            _chartView.setNeedsDisplay()
        }
    
        if (key == "toggleStartZero") {
            _chartView.leftAxis.startAtZeroEnabled = !_chartView.leftAxis.isStartAtZeroEnabled
            _chartView.rightAxis.startAtZeroEnabled = !_chartView.rightAxis.isStartAtZeroEnabled
    
            _chartView.setNeedsDisplay()
        }
    
        if (key == "animateX") {
            _chartView.animate(xAxisDuration: 3.0)
        }
    
        if (key == "animateY") {
            _chartView.animate(yAxisDuration: 3.0)
        }
    
        if (key == "animateXY") {
            _chartView.animate(xAxisDuration: 3.0, yAxisDuration: 3.0)
        }
    
        if (key == "saveToGallery") {
            _chartView.saveToCameraRoll()
        }
    
        if (key == "togglePinchZoom") {
            _chartView.pinchZoomEnabled = !_chartView.isPinchZoomEnabled
    
            _chartView.setNeedsDisplay()
        }
    
        if (key == "toggleAutoScaleMinMax") {
            _chartView.autoScaleMinMaxEnabled = !_chartView.isAutoScaleMinMaxEnabled
            _chartView.notifyDataSetChanged()
        }
    }

    //pragma mark - Actions
    @IBAction func sliderXValueChanged(sender: UISlider) {
        _sliderTextX.text = String(Int(_sliderX.value) + 1)
        
        self.setDataCount(Int(_sliderX.value) + 1, range:Double(_sliderY.value))
    }
    
    @IBAction func sliderValueChanged(sender: UISlider) {
        _sliderTextY.text = String(Int(_sliderY.value))
    
        self.setDataCount(Int(_sliderX.value) + 1, range:Double(_sliderY.value))
    }
}
